### ISSUE TITLE
Users able to see eachothers accounts

### ISSUE THREAD

**Issue by expert-geeks** [pierre-alain-b/rainloop-nextcloud#123]
User A and User B have the same email credentials stored in their nextcloud rainloop settings. (This is a group email account, to be accessed by several staff).
User A opens rainloop and adds a mail account.
User B opens rainloop and can see both the group account and the email account that User A added. User B removes the account that A added, email account is removed for User A also.

I was expecting that the email account details would be associated to the nextcloud user account. It appears that instead, rainloop is using the initial email account as the identifier.. is this the case ?

**Comment by pierre-alain-b** [comment]
I do not understand: how come B could see a user account added by A???

**Comment by expert-geeks** [comment]
If any nextcloud accounts have the same email account listed in 'Additional settings', then add any accounts they add within rainloop will be viewable by the other nextcloud accounts. 

Try it. Add two users, set the same email address & pass in additional settings. Login or impersonate the first user (A). Open rainloop and 'Add account', add an additional email account. Log out.  Log in or impersonate the second user (B) and open rainloop. User B will see the email account that user A added to rainloop.

Is that clearer ?

**Comment by DJCrashdummy** [comment]
i'll try to help out and clear the things up, since i guess to know what @expert-geeks is talking about...

#### general explanation:
well, you have to see rainloop as an "own web-app" because it is intended as a [standalone IMAP-client](https://www.rainloop.net/).  
the nextcloud-app more or less "just" installs an instance of rainloop and integrates it into the nextcloud-GUI (therefore you have to login separately into the rainloop-adminpanel, get the misleading *Warning* mentioned at #62 etc.). the only "real connection" between rainloop and nextcloud is the setting/caching of the credentials (that's why this is the only setting within the nextcloud-GUI) to make the rainloop-app work seamlessly.

#### enough explaining... lets examine your particular case @expert-geeks:
if user A and user B store the same email credentials in nextclouds rainloop-settings (or enter them in the rainloop login form, if they haven't set any working one), then they are indeed logging in into the same rainloop-account! and then it makes sense that both see and can edit/delete the same additional accounts.

#### there are 2 ways (i know) an admin can prevent this:
- either disable the possibility to add additional accounts in the rainloop-adminpanel...  
so every user can add the credentials for the group account or a personal one in the nextcloud-setting and just access this account.
- or choose in nextclouds rainloop-settings to use (working) user/email credentials for all users, if they all have personal accounts (works awesome in combination with the [IMAP-backend](https://apps.nextcloud.com/apps/user_external) :wink: )...  
then every user can add the group account as an additional one to his own rainloop account.

**Comment by pierre-alain-b** [comment]
Thank you, this is much helpful and it does clarify very well the issue.
@expert-geeks with this explanation, do you agree to close the issue?

**Comment by expert-geeks** [comment]
Indeed, that's what I described.. Referring to the mitigations; Unfortunately I need the ability for some users to have more than one account listed, so I can't disable the additional accounts. Also, not everyone has their own email address, hence the need to 'share' email accounts between some users.

If it were possible to edit the directory where the rainloop app saves its settings and have the username or uid appended to it, then settings for that account would be personal to the nextcloud user. I agree this is quite a lot of work and breaks how rainloop currently stores the data, so existing installs would need to be converted at upgrade..

I agree it's a limited problem, but for me this was serious and potentially compromising as I was suddenly presented with users who unexpectedly had access to email accounts and information they should have not been able to see. Luckily this was discovered and mitigated before users could access anything. I suggest a note in the wiki to forewarn other sysadmins wouldn't go amiss.

**Comment by DJCrashdummy** [comment]
> Unfortunately I need the ability for some users to have more than one account listed, so I can't disable the additional accounts. Also, not everyone has their own email address, hence the need to 'share' email accounts between some users.

well what you can do in your unfortunate situation, is...
- either disable additional accounts (and perhaps disable the auto-login),
then all users can/must login into the needed mail account case by case manually...
- or enable additional accounts and still set in the nextcloud-GUI to use user/email credentials for all users for auto-login.
so the users with personal accounts get a convenient auto-login and can add the group-account, and users with no personal account will get to the rainloop-login page because of no (working) user/email credentials. - then they can login into the group-account manually.
the only thing you have to be concerned about is, that no one ads an additional account to the manually logged-in group account!

what i would do either way and hopefully solves you problem, is to restrict the allowed domains via the rainloop-adminpanel:
lets say the nextcloud instance is for one company with one domain and you have a group-account and some personal accounts... then what additional accounts should the users of the group account (without a personal one) add?
**but**, if you have more than one group-account for different groups, then i'm out of ideas... and **the only safe way is to disable additional accounts**.

- - -
> I agree this is quite a lot of work and breaks how rainloop currently stores the data [...]

i'm afraid that this wont get implemented (any time soon)!
- i'm sooooo extremely thankful that @pierre-alain-b still maintains this nextcloud-app :heart: :clap:, and pretty sure this is "a little bit" out of scope.
- and i bet that the rainloop devs don't care about corner cases when using rainloop via nextcloud... at least on github they seem extremely unresponsive. :unamused: 

- - -
> I suggest a note in the wiki to forewarn other sysadmins wouldn't go amiss.

well i'm not sure where to add it... because strictly seen, en-/disabling additional accounts is more a rainloop-related thing; but there it is pretty obvious, if you use the same credentials that you login into the same account. :neutral_face: 
if you've got a good idea where and how to add it, @expert-geeks, then create a PR and i'm pretty sure, (if done well) @pierre-alain-b will merge and appreciate it (beside other users & admins).

**Comment by expert-geeks** [comment]
Thanks for your input. To be honest, I've mitigated the issue by telling users that if they add personal accounts to their work one in rainloop, it'll get shared with users of the shared email account. No-one has done so far. It was more that I want to avoid others from having the same experience.

I appreciate fixing this is more work than anyone has time for. Indeed I'm also very grateful for the work that @pierre-alain-b does keeping Rainloop available for nextcloud users. He's always been friendly and helpful.

All I'm really suggesting is that we add a note in the wiki to the effect that.. The plugin passes the login information from 'additional settings' to the Rainloop app which creates and manages user accounts. Accounts in rainloop are based soley on authenticated email accounts, and do not take account of the nextcloud user which created it. If two or more nextcloud users have the same email account in additional settings, they will view the same 'email account' including any additional email accounts that may have been added subsequently.
